node/bindnode/node.go

```diff
@@ -0,0 +1,1099 @@
+		type goUnion struct {
+			Index int // 0..len(typ.Members)-1
+			Value interface{}
+		}
+		return reflect.TypeOf(goUnion{})
+	}
+	panic(fmt.Sprintf("%T\n", typ))
+}
+
+// Prototype implements an ipld.NodePrototype given a Go pointer type and an
+// IPLD schema type.
+//
+// In this form, it is assumed that the Go type and IPLD schema type are
+// compatible. TODO: check upfront and panic otherwise
+func Prototype(ptrType interface{}, schemaType schema.Type) ipld.NodePrototype {
+	goPtrType := reflect.TypeOf(ptrType)
+	if goPtrType.Kind() != reflect.Ptr {
+		panic("ptrType must be a pointer")
+	}
+	return prototype(goPtrType.Elem(), schemaType)
+}
+
+func prototype(goType reflect.Type, schemaType schema.Type) ipld.NodePrototype {
+	if goType.Kind() == reflect.Invalid {
+		panic("goType must be valid")
+	}
+	if schemaType == nil {
+		panic("schemaType must not be nil")
+	}
+	return &_prototype{schemaType: schemaType, goType: goType}
+}
+
+var (
+	_ ipld.NodePrototype = (*_prototype)(nil)
+
+	_ ipld.Node        = (*_node)(nil)
+	_ schema.TypedNode = (*_node)(nil)
+
+	_ ipld.NodeBuilder   = (*_builder)(nil)
+	_ ipld.NodeAssembler = (*_assembler)(nil)
+
+	_ ipld.MapAssembler = (*_structAssembler)(nil)
+	_ ipld.MapIterator  = (*_structIterator)(nil)
+
+	_ ipld.ListAssembler = (*_listAssembler)(nil)
+	_ ipld.ListIterator  = (*_listIterator)(nil)
+)
+
+type _prototype struct {
+	schemaType schema.Type
+	goType     reflect.Type // non-pointer
+}
+
+func (w *_prototype) NewBuilder() ipld.NodeBuilder {
+	return &_builder{_assembler{
+		schemaType: w.schemaType,
+		val:        reflect.New(w.goType).Elem(),
+	}}
+}
+
+// TODO: consider these Typed interfaces for the schema package
+
+type TypedPrototype interface {
+	ipld.NodePrototype
+
+	Representation() ipld.NodePrototype
+}
+
+type TypedAssembler interface {
+	ipld.NodeAssembler
+
+	Representation() ipld.NodeAssembler
+}
+
+func (w *_prototype) Representation() ipld.NodePrototype {
+	return (*_prototypeRepr)(w)
+}
+
+var (
+	goTypeString = reflect.TypeOf("")
+	goTypeBytes  = reflect.TypeOf([]byte{})
+	goTypeLink   = reflect.TypeOf((*ipld.Link)(nil)).Elem()
+
+	schemaTypeFieldName = schema.SpawnString("fieldNameString")
+)
+
+type _node struct {
+	schemaType schema.Type
+
+	val reflect.Value // non-pointer
+}
+
+// TODO: only expose TypedNode methods if the schema was explicit.
+// type _typedNode struct {
+// 	_node
+// }
+
+func (w *_node) Type() schema.Type {
+	return w.schemaType
+}
+
+func (w *_node) Representation() ipld.Node {
+	return (*_nodeRepr)(w)
+}
+
+func (w *_node) Kind() ipld.Kind {
+	return w.schemaType.TypeKind().ActsLike()
+}
+
+func (w *_node) LookupByString(key string) (ipld.Node, error) {
+	switch typ := w.schemaType.(type) {
+	case *schema.TypeStruct:
+		field := typ.Field(key)
+		if field == nil {
+			return nil, ipld.ErrInvalidKey{
+				TypeName: typ.Name().String(),
+				Key:      basicnode.NewString(key),
+			}
+		}
+		fval := w.val.FieldByName(fieldNameFromSchema(key))
```

Comment:
Maybe. For the time being I think the only way to customize this will be... to bring your own Go type, and change the names to whatever you want, as long as the "shape" of the Go type still matches the schema.